operations/agent_monitor.go

```diff
@@ -0,0 +1,427 @@
+	}
+
+	return client, nil
+}
+
+// createAgentProcess attempts to start an agent subprocess.
+func createAgentProcess(ctx context.Context, m *monitor, retry *retryArgs) (jasper.Process, error) {
+	agentCmdArgs := append([]string{m.clientPath, "agent"}, m.agentArgs...)
+
+	// Copy the monitor's environment to the agent.
+	env := make(map[string]string)
+	for _, envVar := range os.Environ() {
+		keyVal := strings.Split(envVar, "=")
+		env[keyVal[0]] = keyVal[1]
+	}
+
+	var proc jasper.Process
+	var err error
+
+	if err = retryWithArgs(ctx, func() (bool, error) {
+		opts := &jasper.CreateOptions{Environment: env}
+		cmd := m.jasperClient.CreateCommand(ctx).
+			Add(agentCmdArgs).
+			SetOutputSender(level.Info, grip.GetSender()).
+			SetErrorSender(level.Error, grip.GetSender()).
+			ApplyFromOpts(opts).
+			Background(true)
+		if err = cmd.Run(ctx); err != nil {
+			return true, errors.Wrap(err, "failed to create process")
+		}
+
+		jasperIDs := cmd.GetProcIDs()
+
+		// The command should spawn exactly 1 process.
+		if len(jasperIDs) != 1 {
+			catcher := grip.NewBasicCatcher()
+
+			for _, jasperID := range jasperIDs {
+				proc, err = m.jasperClient.Get(ctx, jasperID)
+				if err != nil {
+					catcher.Add(errors.Wrap(err, "could not find erroneous process"))
+					continue
+				}
+				catcher.Add(proc.Signal(ctx, syscall.SIGKILL))
+			}
+
+			return true, errors.Wrapf(catcher.Resolve(), "monitor should spawn exactly 1 process (the agent), but found %d", len(jasperIDs))
+		}
+
+		proc, err = m.jasperClient.Get(ctx, jasperIDs[0])
+		if err != nil {
+			return true, errors.Wrap(err, "could not find agent process")
+		}
+
+		return false, nil
+	}, retry); err != nil {
+		return nil, errors.Wrapf(err, "failed to start agent process '%v'", agentCmdArgs)
+	}
+	return proc, nil
+}
+
+// waitUntilComplete waits until the given process to complete or the context
+// times out. maxDelay determines the time between checks on the process.
+func waitUntilComplete(ctx context.Context, proc jasper.Process, maxDelay time.Duration) (exitCode int, err error) {
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return proc.Wait(ctx)
+		case <-timer.C:
+			if proc.Complete(ctx) {
+				return proc.Wait(ctx)
+			}
+			timer.Reset(maxDelay)
+		}
+	}
+}
+
+// runAgent starts the agent with the necessary args and waits for it to
+// terminate.
+func runAgent(ctx context.Context, m *monitor, retry *retryArgs) error {
+	proc, err := createAgentProcess(ctx, m, retry)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create agent process")
+	}
+
+	exitCode, err := waitUntilComplete(ctx, proc, defaultRetryArgs().maxDelay)
+
+	return errors.Wrapf(err, "agent exited with code %d", exitCode)
+}
+
+// runMonitor runs the monitor loop. It fetches the agent, starts it, and
+// repeats when the agent terminates.
+func runMonitor(ctx context.Context, m *monitor) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
```

Comment:
It might make sense to have some backoff on the continues, in case something goes wrong. We want to avoid DDOSing ourselves.

Reply:
I added a sleep between loops.